### PR TITLE
feat: pinning proxy service

### PR DESCRIPTION
**Pull request by luzzif**
This PR adds an authenticated pinning proxy service to pin data directly to Carrot IPFS nodes. The service implements signature-based web3 authentication with nonces to prevent replay attacks.

The repo structure was overhauled to accommodate for the new package. In particular, a `packages` folder was created with the `pinnin-proxy` service added along with additionally required JS packages and Turbo was added to aid in the build process.

Extensive documentation was added to the new services.

### CODE REVIEW

File: packages/pinning-proxy/src/routes/pins.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    ipfsClusterPassword,
+}: GetPinsRouteParams): ServerRoute => {
+    return {
+        method: "POST",
+        path: "/pins",
+        options: {
+            plugins: {
+                "hapi-swagger": {
+                    payloadType: "form",
+                    description:
```
Comment (guerrap) [inline packages/pinning-proxy/src/routes/pins.ts:22]: Same here for the `description`.

File: packages/pinning-proxy/src/routes/token.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    dbClient,
+    jwtSecretKey,
+}: GetTokenRouteParams): ServerRoute => {
+    return {
+        method: "POST",
+        path: "/token",
+        options: {
+            plugins: {
+                "hapi-swagger": {
+                    description:
```
Comment (guerrap) [inline packages/pinning-proxy/src/routes/token.ts:26]: Ditto for the `description`.

**Review by guerrap — CHANGES_REQUESTED** [review]

File: packages/pinning-proxy/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,52 @@
+FROM node:18-alpine AS base
+
+FROM base AS builder
+RUN apk add --no-cache libc6-compat
+RUN apk update
```
[the remark this answers is not in the document]
Comment (luzzif) [reply packages/pinning-proxy/Dockerfile:5]: Holy, I copied the example given by Turborepo. Let me fix that.

File: packages/pinning-proxy/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+RUN yarn global add turbo
+WORKDIR /app
+COPY . .
+RUN turbo prune --scope=@carrot-kpi/pinning-proxy --docker
+
+FROM base AS installer
+RUN apk add --no-cache libc6-compat
+RUN apk update
```
[the remark this answers is not in the document]
Comment (luzzif) [reply packages/pinning-proxy/Dockerfile:12]: Fixed in 077bde0efc0a66e16e03de3bf94f28561cd34007

File: packages/pinning-proxy/src/db.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import { NONCE_LENGTH_BYTES } from "./commons";
+
+interface InitializeDatabaseParams {
+    connectionString: string;
+}
+
+export const initializeDatabase = async ({
+    connectionString,
+}: InitializeDatabaseParams): Promise<Client> => {
+    const client = await new Client({ connectionString });
```
[the remark this answers is not in the document]
Comment (luzzif) [reply packages/pinning-proxy/src/db.ts]: Fixed in 90b180c55e79336e6d5fefc2650359d2577acef2

File: packages/pinning-proxy/src/routes/login-message.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+export const getLoginMessageRoute = ({
+    dbClient,
+}: GetLoginMessageRouteParams): ServerRoute => {
+    return {
+        method: "PUT",
+        path: "/login-message/{address}",
+        options: {
+            plugins: {
+                "hapi-swagger": {
+                    description:
```
[the remark this answers is not in the document]
Comment (luzzif) [reply packages/pinning-proxy/src/routes/login-message.ts]: Fixed in f2273651c7fc81fed20e9c691878d7d9f2e9ac9b

File: packages/pinning-proxy/src/routes/pins.ts (hunk 1 above)
Comment (luzzif) [reply packages/pinning-proxy/src/routes/pins.ts]: Fixed in f2273651c7fc81fed20e9c691878d7d9f2e9ac9b

File: packages/pinning-proxy/src/routes/token.ts (hunk 2 above)
Comment (luzzif) [reply packages/pinning-proxy/src/routes/token.ts]: Fixed in f2273651c7fc81fed20e9c691878d7d9f2e9ac9b

**Review by guerrap — APPROVED** [review]

**Pull request merged**